packages/query-core/src/types.ts

```diff
@@ -118,6 +132,7 @@ export interface QueryOptions<
   networkMode?: NetworkMode
   gcTime?: number
   queryFn?: QueryFunction<TQueryFnData, TQueryKey, TPageParam>
+  persister?: QueryPersister<TQueryFnData, TQueryKey, TPageParam>
```

Comment:
@emzet93 @TkDodo Just want to move the discussion to a thread, instead of separate comments. (GitLab has that covered 🙄)

> also what do you think about adding the api allowing to restore persisted data to query cache for specific query, so it could be called in advance and allow skipping hard loading state when persisted query is mounted? I think that it should be possible if we pass a persister instance to it, so based on example from this PR it would be sth like:

I believe we could add it. Even better `persister` could be passed as a `default option` for query, so the only thing you would need to specify is `queryKey`.
The issue i see is - how do we skip `queryFn` and `saving to storage` when restoring this way? Cause we only need the first part of `createPersister`?

We could potentially extract `restoring` to a separate function, so you could use only that piece?
Thoughts?

Reply:
'Even better persister could be passed as a default option for query, so the only thing you would need to specify is queryKey.'

Setting persister as default will probably cause every single query to be persisted, right (if not overwritten on query level)? I imagine that it might not always be desired, so we would still need possibility to pass it directly to 'restoreQuery', in case persisted is not defined on defaults.

'We could potentially extract restoring to a separate function, so you could use only that piece?'

That sounds good to me 👍

'I think this would belong to userland. User can get all keys of storage and call multiple times restore function?'

Agree that it should be userland thing. And I had the same idea for calling multiple restore. But @TkDodo made me aware that getting all keys from storage will give us all query hashes, not query keys. Restore should probably takes key, not hash, so it won't help us because I assume that we can convert hash back to key.

Reply:
> We could potentially extract restoring to a separate function, so you could use only that piece? Thoughts?

Yes 👍 

> Setting persister as default will probably cause every single query to be persisted, right (if not overwritten on query level)?

Maybe we could make persister be an optional function? This is a common pattern in RQ, something like:

```
persister: Persister | undefined | ((query: Query) => Persister | undefined)
```

with that, you could define a global persister for all queries that still excludes certain queries, e.g. depending on the `queryKey` or depending on the `meta` field.

Reply:
Yeah that sounds perfect 👍
Two questions though:
- it should still be possible to pass optional `persister` to `queryClient.restoreQuery` in case no default persister is set? Or should setting default persister be mandatory in order to use `restoreQuery`?
- if we set default persister, it still should be possible to disable it for specific query by setting `persister: null/undefined` on `useQuery` level, right?

Reply:
> it should still be possible to pass optional persister to queryClient.restoreQuery in case no default persister is set? Or should setting default persister be mandatory in order to use restoreQuery?

yes this should work.

> if we set default persister, it still should be possible to disable it for specific query by setting persister: null/undefined on useQuery level, right?

yes, the options are shallow merged:

https://github.com/TanStack/query/blob/13b17ee45de3299dd3d3a2dfc3d23ad2f024882a/packages/query-core/src/queryClient.ts#L469-L474

so if you have `persister: undefined`, it would overwrite it.

Reply:
Great! So it sounds like a good way to go.